src/intellisense/notebookPythonPathService.ts

```diff
@@ -0,0 +1,99 @@
+    private _isEnabled: boolean | undefined;
+
+    constructor(
+        @inject(IPythonApiProvider) private readonly apiProvider: IPythonApiProvider,
+        @inject(IVSCodeNotebook) private readonly notebooks: IVSCodeNotebook,
+        @inject(IInteractiveWindowProvider) private readonly interactiveWindowProvider: IInteractiveWindowProvider,
+        @inject(INotebookControllerManager) private readonly notebookControllerManager: INotebookControllerManager,
+        @inject(IInterpreterService) private readonly interpreterService: IInterpreterService,
+        @inject(IConfigurationService) private readonly configService: IConfigurationService
+    ) {}
+
+    public async activate() {
+        if (!this.isPylanceUsingLspNotebooks()) {
+            return;
+        }
+
+        await this.apiProvider.getApi().then((api) => {
+            if (api.registerJupyterPythonPathFunction !== undefined) {
+                api.registerJupyterPythonPathFunction((uri) => this._jupyterPythonPathFunction(uri));
+            }
+        });
+    }
+
+    /**
+    * Returns a boolean indicating whether Pylance's LSP notebooks experiment is enabled.
+    * When this is True, the Python extension starts Pylance for notebooks instead of us.
+    */
+    public isPylanceUsingLspNotebooks() {
+        if (this._isEnabled === undefined) {
+            const isInNotebooksExperiment = this.configService.getSettings().pylanceLspNotebooksEnabled;
+            const pythonVersion = extensions.getExtension(PythonExtension)?.packageJSON.version;
+            const pylanceVersion = extensions.getExtension(PylanceExtension)?.packageJSON.version;
+
+            // Only enable the experiment if we're in the treatment group and the installed
+            // versions of Python and Pylance support the experiment.
+            this._isEnabled =
+                isInNotebooksExperiment &&
+                pythonVersion !== undefined &&
+                semver.satisfies(pythonVersion, '>=2022.6.0 || 2022.5.0-dev') &&
+                pylanceVersion !== undefined &&
+                semver.satisfies(pylanceVersion, '>=2022.4.4-pre.1 || 9999.0.0-dev');
+
+            if (this._isEnabled) {
+                traceInfo('Pylance LSP Notebooks experiment is enabled.');
+            }
+        }
+
+        return this._isEnabled;
+    }
+
+    /**
+    * Called by the Python extension when Pylance needs the python.exe path for a notebook.
+    */
+    private async _jupyterPythonPathFunction(uri: Uri): Promise<string | undefined> {
+        const notebook = findAssociatedNotebookDocument(uri, this.notebooks, this.interactiveWindowProvider);
+        const controller = notebook
+            ? this.notebookControllerManager.getSelectedNotebookController(notebook)
+            : undefined;
+        const interpreter = controller
+            ? controller.connection.interpreter
+            : await this.interpreterService.getActiveInterpreter(uri);
```

Comment:
I think we should probably let python decide the active interpreter unless a notebook was found for the URI. I'm not sure how this function is called, but if it's called for other files, we'll always be returning the active interpreter. I think it makes more sense to only do that notebooks.

Maybe the ```if (!interpreter)``` on line 89 could be changed to always return undefined if we don't find a notebook.

Reply:
My intent was to only call this for notebooks. Currently I have Python only calling this function when it gets config requests for paths ending in `.ipynb`. Would it be better to always call into Jupyter and let it figure out if the file is a notebook?

Is Jupyter currently activated in non-notebook scenarios? If not, would we want to avoid that?

Reply:
I think Jupyter would better know if the file is a notebook or not.

Juptyer is activated in non notebook scenarios for this use case:
https://code.visualstudio.com/docs/python/jupyter-support-py

Reply:
If jupyter figures it out, it can also find interactive window requests.